### ISSUE TITLE
How to change SI unit when exporting?

### ISSUE THREAD

**Issue by gkovacsds** [IfcOpenShell/IfcOpenShell#2186]
Hello, 
when using the C++ export API, how to change the length unit to Centimeter instead of Millimeter?

I tried this
file.addEntity(new IfcSchema::IfcSIUnit(IfcSchema::IfcUnitEnum::IfcUnit_LENGTHUNIT, IfcSchema::IfcSIPrefix::IfcSIPrefix_CENTI,
		IfcSchema::IfcSIUnitName::IfcSIUnitName_METRE));
but it just creates duplicate entries.

I could not manage it with accessing
project->UnitsInContext();
I don't know how to access and change the contained unit elements in it.
Could you give me a code example?

Thank you.

**Comment by Moult** [comment]
At a blind guess, `project->UnitsInContext()->setUnits(units)` might be what you want? Where `units` is `Schema::IfcUnit::list::ptr items = project->UnitsInContext()->Units();` , remove the old unit, then `units->push(new_unit)`.

As a general question, are there C++ code examples outside the examples dir? Perhaps this might be a worthy goal of the new site and the new docs, to have a clear C++ hello world starter on the main site and perhaps a series of useful C++ snippets directly in the Sphinx docs?

**Comment by aothms** [comment]
The following code should work, assuming you have a context with an existing SI unit for length.

~~~C++
IfcHierarchyHelper<Ifc2x3> h;
auto proj = h.getSingle<Ifc2x3::IfcProject>();
auto units = proj->UnitsInContext()->Units();
for (auto& u : *units) {
	Ifc2x3::IfcSIUnit* nu;
	if (nu = u->as<Ifc2x3::IfcSIUnit>()) {
		if (nu->UnitType() == Ifc2x3::IfcUnitEnum::IfcUnit_LENGTHUNIT) {
			nu->setPrefix(Ifc2x3::IfcSIPrefix::IfcSIPrefix_CENTI);
		}
	}
}
~~~

IFC units are indeed notoriously painful in a strictly typed language due to all the selects and subtyping.

**Comment by gkovacsds** [comment]
Thank you, it works fine.
One more question though - how do I set METER as length unit?
`nu->setPrefix(0);`
throws an exception, and there is - strangely - no 10^0 valued constant in IfcSIPrefix enum type.

**Comment by aothms** [comment]
The function is:

    void IfcSIUnit::setPrefix(boost::optional<::Ifc4::IfcSIPrefix::Value> v);

You can use `nu->setPrefix(boost::none);`

**Comment by gkovacsds** [comment]
Have you tried this? It throws an exception for me. I'm using IFC4 schema.
Basically it equals to 
`nu->setPrefix(0);`

As solution, I just use text search and replace post-processing at the moment if I need meters.

**Comment by aothms** [comment]
You're correct, this wasn't correct handled (setters with optional attrs are new like this in 0.7). Should be fixed now.

**Comment by Moult** [comment]
Bump @gkovacsds can you confirm the fix? :)

**Comment by gkovacsds** [comment]
Actually I did not try it for lack of time. I close the issue though, thank you.
